sematic/log_reader.py

```diff
@@ -422,40 +499,69 @@ def line_stream_from_log_directory(
         ),
+        reverse=reverse,
     )
-    found_cursor_file = cursor_file is None
-    found_cursor_line = cursor_line_index is None
-    for log_file in log_files:
-        if log_file == cursor_file:
-            found_cursor_file = True
-        if not found_cursor_file:
-            continue
+    index_of_log_file = 0
+    if cursor_file is not None:
+        if cursor_file not in log_files:
+            raise ValueError(f"Cursor referenced invalid log file {cursor_file}.")
+        index_of_log_file = log_files.index(cursor_file)
+
+    cursor_file = log_files[index_of_log_file]
+    log_files_to_traverse = log_files[index_of_log_file:]
+    for log_file in log_files_to_traverse:
+        start_line_index = sys.maxsize if reverse else 0
+        if log_file == cursor_file and cursor_line_index is not None:
+            start_line_index = cursor_line_index
+
         text_stream: Iterable[str] = storage.get_line_stream(log_file)
-        for i_line, line in enumerate(text_stream):
-            if (
-                (not found_cursor_line)
-                and cursor_line_index is not None
-                and i_line < cursor_line_index
-            ):
+        yield from _stream_from_text_stream_from_index(
+            log_file, start_line_index, reverse, text_stream
+        )
+
+
+def _stream_from_text_stream_from_index(
+    log_file: str, start_line_index: int, reverse: bool, text_stream: Iterable[str]
+) -> Iterable[LogLine]:
+    """Starting at the given index within the file, stream logs in desired direction.
+
+    When going forward, the line pointed to by the index will be included. When going in
+    the reverse direction, the line pointed to by the index will be omitted.
+    """
+    # For purposes of this function, "preceding" refers to indices
+    # lower than the start index, regardless of traversal direction.
+    lines_preceding_cursor: List[LogLine] = []
+    for i_line, line in enumerate(text_stream):
+        as_log_line = LogLine(
+            source_file=log_file,
+            source_file_index=i_line,
+            line=line,
+        )
+        if i_line < start_line_index:
+            if reverse:
+                lines_preceding_cursor.append(as_log_line)
+            else:
                 continue
-            found_cursor_line = True
-            yield LogLine(
-                source_file=log_file,
-                source_file_index=i_line,
-                line=line,
-            )
-        if cursor_file is not None and log_file >= cursor_file:
-            # we automatically know we hit the cursor line if we are at the end
-            # of the cursor file or in a file that comes after it.
-            found_cursor_line = True
+        elif reverse:
+            # once we hit the cursor, we have everything we
+            # need to go backwards
+            break
+        else:
+            yield as_log_line
+    if reverse:
+        yield from reversed(lines_preceding_cursor)
```

Comment:
This function contains 3 `if reverse` statements.
Please consider refactoring it to have only one of these statements at the top, then 2 less complicated and more readable branches, even if the for statement is duplicated.

Reply:
done